File: java/bob/src/main/java/Bob.java
import java.util.Locale;

public class Bob {
    public String hey(String speech) {
        if (isEmpty(speech)) return "Fine. Be that way!";
        if (isQuestion(speech) && isYelling(speech)) {
            return "Calm down, I know what I'm doing!";
        }
        if (isQuestion(speech)) return "Sure.";
        if (isYelling(speech)) return "Whoa, chill out!";
        return "Whatever.";
    }

    private boolean isYelling(String speech) {
        String yelledSpeech = speech.toUpperCase(Locale.ROOT);
        String spokenSpeech = speech.toLowerCase(Locale.ROOT);
        return speech.equals(yelledSpeech) && !speech.equals(spokenSpeech);
    }

    private boolean isQuestion(String speech) {
        return speech.replaceAll("\\s+","").endsWith("?");
    }

    private boolean isEmpty(String speech) {
        return speech.replaceAll("\\s+","").length() == 0;
    }
}

File: go/bob/bob.go
// This is a "stub" file.  It's a little start on your solution.
// It's not a complete solution though; you have to write some code.

// Package bob should have a package comment that summarizes what it's about.
// https://golang.org/doc/effective_go.html#commentary
package bob

import (
	"regexp"
	"strings"
	"unicode/utf8"
)

var ws = regexp.MustCompile(`^[\s\t\n\r]*$`)
var hl = regexp.MustCompile(`[a-zA-Z]+`)

// Hey should have a comment documenting it.
func Hey(remark string) string {
	switch {
	case ws.MatchString(remark):
		return "Fine. Be that way!"
	case isQuestion(remark) && isYell(remark):
		return "Calm down, I know what I'm doing!"
	case isQuestion(remark):
		return "Sure."
	case isYell(remark):
		return "Whoa, chill out!"
	default:
		return "Whatever."
	}
}

func isQuestion(s string) bool {
	s = strings.TrimSpace(s)
	return s[utf8.RuneCountInString(s)-1] == '?'
}

func isYell(s string) bool {
	return hl.MatchString(s) && s == strings.ToUpper(s)
}
